File: Python/ByRound/1926/1926_F_Vlad_and_Avoiding_X.py
# Problem: CF 1926 F - Vlad and Avoiding X
# https://codeforces.com/contest/1926/problem/F

"""
Algorithm: Dynamic Programming with Memoization (DFS)
Approach: 
- The problem is to eliminate all black cells that have all four diagonal neighbors also black.
- This is done by trying all combinations of flipping cells in a bipartite manner (based on parity of (i+j)).
- States are represented as bitmasks of which cells are black.
- For each state, we determine the minimum number of flips needed by recursively trying to resolve conflicts.
- Time Complexity: O(2^49) which is too large for direct computation, but due to pruning from memoization and small grid, it's manageable.
Space Complexity: O(2^49) in worst case for memoization.
"""

import sys

input = lambda: sys.stdin.readline().rstrip()
memo = {}


def solve():
    s = []
    for _ in range(7):
        s.append(input())

    def dfs(state, parity):
        # Base case: no more conflicts
        if state == 0:
            return 0
        # Memoization
        if (state, parity) in memo:
            return memo[(state, parity)]
        # Collect all conflicting cells based on parity
        t = []
        for i in range(7):
            for j in range(7):
                if (i + j) % 2 == parity:
                    data = [0, 0]
                    # Check for all diagonal neighbors including current
                    for x, y in [
                        (i - 1, j - 1),
                        (i - 1, j + 1),
                        (i + 1, j - 1),
                        (i + 1, j + 1),
                        (i, j),
                    ]:
                        # If in bounds and black cell
                        if 0 <= 7 * x + y < 49 and state & 1 << (7 * x + y):
                            data[0] += 1
                            data[1] |= 1 << (7 * x + y)
                    if data[0] > 0:
                        t.append(data.copy())
        # Sort to try more impactful moves first
        t.sort(reverse=True)
        res = 212  # Large number to represent infinity
        for cnt, cancel in t:
            # Try flipping one of the conflicting cells
            res = min(res, 1 + dfs(state ^ cancel, parity))
        memo[(state, parity)] = res
        return res

    # Build initial state (bitmask for each parity group)
    state = [0, 0]
    for i in range(1, 6):
        for j in range(1, 6):
            # If it's a black cell with all four diagonal neighbors also black
            if (
                s[i][j] == "B"
                and s[i - 1][j - 1] == "B"
                and s[i + 1][j - 1] == "B"
                and s[i + 1][j + 1] == "B"
                and s[i - 1][j + 1] == "B"
            ):
                state[(i + j) % 2] |= 1 << (7 * i + j)
    
    # Calculate minimal operations for both parity groups
    res = dfs(state[0], 0) + dfs(state[1], 1)
    print(res)
    return


T = int(input())
for _ in range(T):
    solve()


# https://github.com/VaHiX/CodeForces/

File: CPP/ByRound/1926/1926_F_Vlad_and_Avoiding_X.cpp
// Problem: CF 1926 F - Vlad and Avoiding X
// https://codeforces.com/contest/1926/problem/F

/*
Algorithm: Backtracking with memoization
Time Complexity: O(2^k) where k is the number of black cells that form the forbidden pattern
Space Complexity: O(1) as we use fixed size 7x7 grid and recursion stack depth is bounded

Approach:
1. Identify all black cells that have all four diagonal neighbors also black (forbidden pattern)
2. Split these cells into two groups based on (i+j) % 2 to avoid conflicts
3. Use backtracking to find minimum operations by trying to flip each cell in groups
4. For each forbidden pattern, try flipping any of the 4 involved cells to break the pattern
5. Return minimum operations required to eliminate all forbidden patterns
*/

#include <ext/pb_ds/assoc_container.hpp>
#include <ext/pb_ds/tree_policy.hpp>
#include <stddef.h>
#include <ext/pb_ds/tag_and_trait.hpp>
#include <iostream>
#include <algorithm>
#include <bitset>
#include <cmath>
#include <functional>
#include <iterator>
#include <queue>
#include <stack>
#include <string>
#include <utility>
#include <vector>

#pragma GCC target("avx2")
#pragma GCC optimize("O3")
#pragma GCC optimize("Ofast")
#pragma GCC optimize("unroll-loops")
#pragma GCC optimization("O3")
#pragma GCC optimization("tree-vectorize")

#define int long long
#define INF 1000000007
#define PI 3.14159265358979323846
#define cinarr(arr, n) fr(i, 0, n) cin >> arr[i];
#define coutarr(arr, n) fr(i, 0, n) cout << arr[i] << " ";
#define vall(arr) arr.begin(), arr.end()

using namespace std;

// Helper function to find if a cell is part of forbidden pattern
bool isForbidden(int i, int j, vector<vector<char>>& arr) {
    return (arr[i][j] == 'B' && arr[i - 1][j - 1] == 'B' &&
            arr[i - 1][j + 1] == 'B' && arr[i + 1][j - 1] == 'B' &&
            arr[i + 1][j + 1] == 'B');
}

// Recursive backtracking function to find minimum operations
int f(vector<pair<int, int>>& v, int ind, vector<vector<char>>& arr) {
    // Base case: no more forbidden patterns to process
    if (ind >= v.size())
        return 0;
    
    int i = v[ind].first;
    int j = v[ind].second;
    
    // If current pattern is not present, skip this cell
    if (!isForbidden(i, j, arr)) {
        return f(v, ind + 1, arr);
    }
    
    int ans = INF;
    
    // Try flipping the center cell
    arr[i][j] = 'W';
    ans = min(ans, 1 + f(v, ind + 1, arr));
    arr[i][j] = 'B';
    
    // Try flipping the top-right cell
    arr[i - 1][j + 1] = 'W';
    ans = min(ans, 1 + f(v, ind + 1, arr));
    arr[i - 1][j + 1] = 'B';
    
    // Try flipping the bottom-left cell
    arr[i + 1][j - 1] = 'W';
    ans = min(ans, 1 + f(v, ind + 1, arr));
    arr[i + 1][j - 1] = 'B';
    
    // Try flipping the bottom-right cell
    arr[i + 1][j + 1] = 'W';
    ans = min(ans, 1 + f(v, ind + 1, arr));
    arr[i + 1][j + 1] = 'B';
    
    return ans;
}

signed main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    cout.tie(0);
    
    int tt;
    cin >> tt;
    while (tt--) {
        int n = 7;
        vector<vector<char>> arr(7, vector<char>(7));
        
        // Read input grid
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                cin >> arr[i][j];
            }
        }
        
        vector<pair<int, int>> v, v2; // Split forbidden cells into two groups
        
        // Identify forbidden patterns and group them
        for (int i = 1; i < n - 1; i++) {
            for (int j = 1; j < n - 1; j++) {
                if (isForbidden(i, j, arr)) {
                    // Group cells based on parity of (i+j) to avoid conflicts
                    if ((i + j) % 2)
                        v.push_back({i, j});
                    else
                        v2.push_back({i, j});
                }
            }
        }
        
        // Calculate minimum operations for both groups
        cout << f(v, 0, arr) + f(v2, 0, arr) << '\n';
    }
    return 0;
}


// https://github.com/VaHiX/CodeForces/